File: exam-system-java/server/src/main/java/com/yyjy/exam/paper/api/PaperController.java
package com.yyjy.exam.paper.api;

import cn.dev33.satoken.annotation.SaCheckRole;
import com.yyjy.exam.common.convention.result.R;
import com.yyjy.exam.entity.paper.dto.PaperAiSaveDto;
import com.yyjy.exam.entity.paper.dto.PaperDetail;
import com.yyjy.exam.entity.paper.dto.PaperListQuery;
import com.yyjy.exam.entity.paper.entity.Paper;
import com.yyjy.exam.entity.paper.entity.PaperCategoriesFetcher;
import com.yyjy.exam.entity.paper.entity.PaperFetcher;
import com.yyjy.exam.entity.paper.io.req.PaperQuestionAddReq;
import com.yyjy.exam.entity.paper.io.req.PaperSaveInput;
import com.yyjy.exam.entity.paper.io.req.PaperUpdateInput;
import com.yyjy.exam.paper.service.PaperService;
import lombok.RequiredArgsConstructor;
import org.babyfish.jimmer.client.FetchBy;
import org.babyfish.jimmer.sql.fetcher.Fetcher;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@SaCheckRole("admin")
@RestController
@RequestMapping("/paper")
@RequiredArgsConstructor
public class PaperController {
	
	public static final Fetcher<Paper> PAPER_ITEM =
			PaperFetcher.$
					.name()
					.description()
					.category(PaperCategoriesFetcher.$.name())
					.published()
					.totalScore()
					.questionCount()
					.duration()
					.start()
					.end()
					.createTime();
	
	private final PaperService paperService;
	
	
	/**
	 * 更新试卷
	 */
	@PutMapping("/update")
	public R<Void> updatePaper(@RequestBody PaperUpdateInput paper) {
		paperService.updatePaper(paper);
		return R.ok();
	}
	
	/**
	 * 删除试卷
	 */
	@DeleteMapping("/remove/{id}")
	public R<Void> removePaper(@PathVariable int id) {
		paperService.removePaper(id);
		return R.ok();
	}
	
	/**
	 * 获取试卷详情
	 */
	@GetMapping("/{id}")
	public R<PaperDetail> getPaper(@PathVariable int id) {
		return R.ok(paperService.getPaper(id));
	}
	
	/**
	 * 获取试卷列表
	 */
	@GetMapping("/list")
	public R<List<@FetchBy("PAPER_ITEM") Paper>> listPapers(PaperListQuery query) {
		
		return R.ok(paperService.listPapersByNameAndStatus(query, PAPER_ITEM));
	}
	
	/**
	 * 新增试卷 v2（不含题目，返回试卷ID）
	 */
	@PostMapping("/v2/add")
	public R<Integer> addPaperV2(@RequestBody PaperSaveInput paper) {
		int id = paperService.addPaper(paper);
		return R.ok(id);
	}
	
	/**
	 * 发布试卷（PUBLISHED false -> true）
	 */
	@PostMapping("/{id}/publish")
	public R<Void> publishPaper(@PathVariable int id) {
		paperService.publishPaper(id);
		return R.ok();
	}
	
	/**
	 * 获取试卷题目列表
	 */
	@GetMapping("/{id}/questions")
	public R<List<PaperDetail.TargetOf_questions>> getPaperQuestions(@PathVariable int id) {
		return R.ok(paperService.getPaperQuestions(id));
	}
	
	/**
	 * 批量添加试卷题目（upsert）
	 */
	@PostMapping("/{id}/questions")
	public R<Void> addPaperQuestions(@PathVariable int id, @RequestBody PaperQuestionAddReq input) {
		paperService.addPaperQuestions(id, input);
		return R.ok();
	}
	
	/**
	 * 删除试卷中的一道题目
	 */
	@DeleteMapping("/{id}/questions/{questionId}")
	public R<Void> removePaperQuestion(@PathVariable int id, @PathVariable long questionId) {
		paperService.removePaperQuestion(id, questionId);
		return R.ok();
	}
	
	/**
	 * 更新试卷中题目的分数
	 */
	@PutMapping("/{id}/questions/{questionId}/score")
	public R<Void> updatePaperQuestionScore(
			@PathVariable int id,
			@PathVariable long questionId,
			@RequestParam int score
	) {
		paperService.updatePaperQuestionScore(id, questionId, score);
		return R.ok();
	}
	
	/**
	 * AI智能组卷
	 */
	@PostMapping("/ai")
	public R<Paper> aiPaper(@RequestBody PaperAiSaveDto dto) {
		return R.ok(paperService.aiCreatePaper(dto));
	}
}

File: exam-system-server/src/main/kotlin/com/yyjy/controller/PaperController.kt
package com.yyjy.controller

import com.yyjy.common.R
import com.yyjy.constants.PaperStatus
import com.yyjy.models.dto.PaperAiSaveDto
import com.yyjy.models.entity.*
import com.yyjy.models.entity.dto.PaperDetail
import com.yyjy.models.entity.dto.PaperSaveInput
import com.yyjy.models.entity.dto.PaperUpdateInput
import com.yyjy.service.PaperService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import org.babyfish.jimmer.client.FetchBy
import org.babyfish.jimmer.client.meta.Api
import org.babyfish.jimmer.sql.kt.ast.expression.case
import org.babyfish.jimmer.sql.kt.ast.expression.eq
import org.babyfish.jimmer.sql.kt.fetcher.newFetcher
import org.springframework.web.bind.annotation.*


@Api
@Tag(name = "试卷模块")
@RequestMapping("/paper")
@RestController
class PaperController(
    private val paperService: PaperService,
) {

    @Api
    @Operation(summary = "新增试卷")
    @PostMapping("/add")
    fun addPaper(@RequestBody paper: PaperSaveInput): R<String?> {
        paperService.addPaper(paper)
        return R.ok()
    }

    @Api
    @Operation(summary = "更新试卷")
    @PutMapping("/update")
    fun updatePaper(@RequestBody paper: PaperUpdateInput): R<String?> {
        paperService.updatePaper(paper)
        return R.ok()
    }

    @Api
    @Operation(summary = "删除试卷")
    @DeleteMapping("/remove/{id}")
    fun removePaper(@PathVariable id: Int): R<String?> {
        paperService.removePaper(id)
        return R.ok()
    }

    @Api
    @Operation(summary = "获取试卷详情")
    @GetMapping("/{id}")
    fun getPaper(@PathVariable id: Int): R<PaperDetail> {
        return R.ok(paperService.getPaper(id))
    }

    @Api
    @Operation(summary = "试卷列表")
    @GetMapping("/list")
    fun listPapers(
        @RequestParam(name = "name", required = false) name: String?, // 明确可选
        @RequestParam(name = "status", required = false) status: PaperStatus?
    ): R<List<@FetchBy("PAPER_ITEM") Paper>> {
        return R.ok(paperService.listPapersByNameAndStatus(name, status, PAPER_ITEM))
    }

    @Api
    @Operation(summary = "AI智能组卷", description = "根据用户输入的规则，智能生成试卷，包括选择题、判断题、简答题等多种题型。")
    @PostMapping("/ai")
    fun aiPaper(@RequestBody paperAiSaveDto: PaperAiSaveDto): R<Paper> {
        return R.ok(paperService.aiCreatePaper(paperAiSaveDto))
    }

    @Api
    @PostMapping("/{id}/status") // 处理POST请求
    @Operation(summary = "更新试卷状态", description = "修改试卷状态：发布试卷供学生考试或停止试卷禁止考试") // API描述
    fun updatePaperStatus(
        @PathVariable id: Int,
        status: String
    ): R<String?> {
        paperService.updateStatus(id, status)
        return R.ok()
    }


    companion object {
        val PAPER_ITEM = newFetcher(Paper::class).by {
            name()
            description()
            status()
            totalScore()
            questionCount()
            duration()
            createTime()
        }
        val PAPER_DETAIL = newFetcher(Paper::class).by {
            PAPER_ITEM
            paperQuestions(
                childFetcher = newFetcher(PaperQuestion::class).by {
                    question(
                        newFetcher(Questions::class).by {
                            title()
                            type()
                            score()
                            questionChoices(
                                newFetcher(QuestionChoices::class).by {
                                    content()
                                    correct()
                                }
                            )
                        },

                    )
                },
                // 对题目进行排序
                cfgBlock = {
                    filter {
                        orderBy(
//                            table.question.id.desc()
                                    case()
                                        .match(table.question.type.eq("CHOICE"), 1)
                                        .match(table.question.type.eq("JUDGE"), 2)
                                        .match(table.question.type.eq("TEXT"), 3)
                                        .otherwise(99)
                        )
                    }
                }
            )
        }
    }
}
